### ISSUE TITLE
Feedback request: give semantic recompilations in Elixir master a try

### ISSUE THREAD

**Issue by josevalim** [elixir-lang/elixir#11203]
Hi everyone, #11201 has been merged. Given it does large changes to how the compiler tracks recompilations across dependencies, I would really appreciate if folks would give Elixir master a try. When doing so, you should watch out for if parts of the project does not recompile whenever you add, remove, update, or configure a dependency. This will most likely happen when switching between branches. If you ever have to `rm -rf _build` to fix things, then it means something is wrong.

Here is the relevant changelog notice about the new features:

### Semantic recompilation

Elixir v1.13 comes with many improvements to the compiler, so it recompiles your files less frequently. In particular:

  * The digest of the files are considered in addition to their size. This avoids recompiling many files when switching or rebasing branches.

  * Changing your `mix.exs` will no longer trigger a full recompilation, unless you specifically change the configurations used by the Elixir compiler (`:elixirc_paths` and `:elixirc_options`).

  * Changing compile-time configuration files (`config/config.exs` and any other file imported from it) now only recompiles the project files that depend on the reconfigured applications, instead of a full recompilation. However, if you change the configuration of your application itself, the whole project is still recompiled.

  * Adding or updating a dependency now only recompiles the project files that depend on the modified a dependency. Removing a dependency still triggers a whole project recompilation.

  * If your project has both Erlang and Elixir files, changing an Erlang file will now recompile only the Elixir files that depend on it.

In a nutshell, Elixir went from triggering full recompilations whenever any of `mix.exs`, `config/config.exs`, `src/*`, and `mix.lock` changed on disk to semantic recompilations where it only fully recompiles when:

  * you change the compilation options in `mix.exs`
  * you change the configuration for the current project in `config/config.exs`
  * you remove a dependency

/cc @marcandre @ericmj @sasa1977

**Comment by crbelaus** [comment]
I've been using Elixir master in my day job project since I saw this issue yesterday. The project is 4 years old and has around 1200 Elixir files, so I think that it is a good test case for the recompilations.

Using Elixir master was super easy. I simply had to `asdf install elixir master` and then `asdf local elixir master`. The project is running without requiring any changes to the code. There are just a few warnings about soon-to-be deprecated function calls that we will have to update before making the Elixir upgrade definitive.

Regarding the topic of this issue: I've seen an enormous performance improvement when switching branches and rebasing. Previously switching branches used to trigger big compilations, but now recompilations are really low as expected.
I didn't experience any hiccup nor had to `rm -rf _build/` either. **Everything is working perfectly for now.**

On the other hand I am planning to update our CI to use the latest Elixir version and leverage the new `mix xref graph --label compile-connected --fail-above` to keep out transitive compile dependencies under control.
This has been a traditional problem in our codebase for the last years and improvements in this area are really welcome.

**Comment by juanperi** [comment]
Is the code reloader from phoenix supposed to be updated as well to test this out?
The following workflow is generating problems for me:
```
mix phx.server
mix deps.clean XXX
mix deps.install
mix compile
mix pxh.server
```
and then I got the following error
```
** (RuntimeError) could not compile application: sylar.

You must restart your server after changing the following config or lib files:

  * _build/dev/lib/sylar/.mix/compile.lock


    (phoenix 1.5.11) lib/phoenix/code_reloader/server.ex:221: Phoenix.CodeReloader.Server.mix_compile_unless_stale_config/2
    (phoenix 1.5.11) lib/phoenix/code_reloader/server.ex:176: Phoenix.CodeReloader.Server.mix_compile/3
    (phoenix 1.5.11) lib/phoenix/code_reloader/server.ex:75: anonymous fn/2 in Phoenix.CodeReloader.Server.handle_call/3
    (phoenix 1.5.11) lib/phoenix/code_reloader/server.ex:289: Phoenix.CodeReloader.Server.proxy_io/1
    (phoenix 1.5.11) lib/phoenix/code_reloader/server.ex:73: Phoenix.CodeReloader.Server.handle_call/3
    (stdlib 3.15.1) gen_server.erl:721: :gen_server.try_handle_call/4
    (stdlib 3.15.1) gen_server.erl:750: :gen_server.handle_msg/6
    (stdlib 3.15.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
```
I seem to remember reading one of the issues about reverting something related to this in this past few days.
Let me know if you want me to try to reproduce this with a smaller app, or if it's not supposed to work yet

**Comment by josevalim** [comment]
Unfortunately for Phoenix you need to restart the server because it won’t pick up the new dependencies automatically. But restarting it should be faster than before too.

**Comment by juanperi** [comment]
right, that's what i'm doing. But that's the issue. Even when restarting, it doesn't recompile. And in this case, I'm executing `mix compile` before restarting the server
the steps are:
```
mix phx.server
# stop server
mix deps.clean XXX
mix deps.install
mix compile
mix pxh.server
```
So, phoenix doesn't start after the last step. I need to go to the config/config and update one of the configs of my app to kick the recompilation of the whole project so phoenix can start

**Comment by juanperi** [comment]
I was just able to replicate it with a vanilla phoenix app. Steps:
```
mix phx.new the_phoenix_bug --no-ecto --no-webpack
cd the_phoenix_bug
# edit asdf tool-versions to use elixir master
mix phx.server
mix deps.clean gettext
mix deps.get
mix phx.server
```
after that, when visiting the app i see:
```
$ mix phx.server
warning: use Mix.Config is deprecated. Use the Config module instead
  config/config.exs:8

warning: use Mix.Config is deprecated. Use the Config module instead
  config/dev.exs:1

==> gettext
Compiling 1 file (.yrl)
Compiling 1 file (.erl)
Compiling 20 files (.ex)
Generated gettext app
==> the_phoenix_bug
Generated the_phoenix_bug app
[info] Running ThePhoenixBugWeb.Endpoint with cowboy 2.9.0 at 0.0.0.0:4000 (http)
[info] Access ThePhoenixBugWeb.Endpoint at http://localhost:4000
** (RuntimeError) could not compile application: the_phoenix_bug.

You must restart your server after changing the following config or lib files:

  * _build/dev/lib/the_phoenix_bug/.mix/compile.lock


    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:221: Phoenix.CodeReloader.Server.mix_compile_unless_stale_config/2
    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:176: Phoenix.CodeReloader.Server.mix_compile/3
    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:75: anonymous fn/2 in Phoenix.CodeReloader.Server.handle_call/3
    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:289: Phoenix.CodeReloader.Server.proxy_io/1
    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:73: Phoenix.CodeReloader.Server.handle_call/3
    (stdlib 3.15.1) gen_server.erl:721: :gen_server.try_handle_call/4
    (stdlib 3.15.1) gen_server.erl:750: :gen_server.handle_msg/6
    (stdlib 3.15.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3

** (RuntimeError) could not compile application: the_phoenix_bug.

You must restart your server after changing the following config or lib files:

  * _build/dev/lib/the_phoenix_bug/.mix/compile.lock


    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:221: Phoenix.CodeReloader.Server.mix_compile_unless_stale_config/2
    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:176: Phoenix.CodeReloader.Server.mix_compile/3
    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:75: anonymous fn/2 in Phoenix.CodeReloader.Server.handle_call/3
    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:289: Phoenix.CodeReloader.Server.proxy_io/1
    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:73: Phoenix.CodeReloader.Server.handle_call/3
    (stdlib 3.15.1) gen_server.erl:721: :gen_server.try_handle_call/4
    (stdlib 3.15.1) gen_server.erl:750: :gen_server.handle_msg/6
    (stdlib 3.15.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3

** (RuntimeError) could not compile application: the_phoenix_bug.

You must restart your server after changing the following config or lib files:

  * _build/dev/lib/the_phoenix_bug/.mix/compile.lock


    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:221: Phoenix.CodeReloader.Server.mix_compile_unless_stale_config/2
    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:176: Phoenix.CodeReloader.Server.mix_compile/3
    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:75: anonymous fn/2 in Phoenix.CodeReloader.Server.handle_call/3
    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:289: Phoenix.CodeReloader.Server.proxy_io/1
    (phoenix 1.5.12) lib/phoenix/code_reloader/server.ex:73: Phoenix.CodeReloader.Server.handle_call/3
    (stdlib 3.15.1) gen_server.erl:721: :gen_server.try_handle_call/4
    (stdlib 3.15.1) gen_server.erl:750: :gen_server.handle_msg/6
    (stdlib 3.15.1) proc_lib.erl:226: :proc_lib.init_p_do_apply/3
```

**Comment by josevalim** [comment]
Thanks. I know exactly what is wrong. I will send a PR to Elixir soon.

**Comment by josevalim** [comment]
Thank you @juanperi. It has been fixed in Elixir master, please update. :)

**Comment by josevalim** [comment]
Hi everyone! It seems everything is going well, so I will go ahead and close this. If you stumble into anything, feel free to open an issue!
